### PR TITLE
WIP - experiment: try mage

### PR DESCRIPTION
**Pull request by n0izn0iz**
# Examples

## target with a dep

Screenshot from 2020-12-07 23-42-17

## building an already up-to-date target

Screenshot from 2020-12-08 00-18-33

## another target

```
❯ mage xcodeGen
🔨 XcodeGen: building
🤖 . 🏠 /tmp/xcodegen_dl js/ios/vendor/xcodegen
🌏 . ⬇️  https://github.com/yonaskolb/XcodeGen/archive/2.18.0.zip > /tmp/xcodegen_dl/xcodegen.zip
🏃 . ❯ unzip -o -q /tmp/xcodegen_dl/xcodegen.zip -d /tmp/xcodegen_dl
🏃 . ❯ cp -rf /tmp/xcodegen_dl/XcodeGen-2.18.0/.gitattributes /tmp/xcodegen_dl/XcodeGen-2.18.0/.github /tmp/xcodegen_dl/XcodeGen-2.18.0/.gitignore /tmp/xcodegen_dl/XcodeGen-2.18.0/.swiftformat /tmp/xcodegen_dl/XcodeGen-2.18.0/Assets /tmp/xcodegen_dl/XcodeGen-2.18.0/CHANGELOG.md /tmp/xcodegen_dl/XcodeGen-2.18.0/CONTRIBUTING.md /tmp/xcodegen_dl/XcodeGen-2.18.0/Docs /tmp/xcodegen_dl/XcodeGen-2.18.0/LICENSE /tmp/xcodegen_dl/XcodeGen-2.18.0/Makefile /tmp/xcodegen_dl/XcodeGen-2.18.0/Package.resolved /tmp/xcodegen_dl/XcodeGen-2.18.0/Package.resources /tmp/xcodegen_dl/XcodeGen-2.18.0/Package.swift /tmp/xcodegen_dl/XcodeGen-2.18.0/README.md /tmp/xcodegen_dl/XcodeGen-2.18.0/RELEASE.md /tmp/xcodegen_dl/XcodeGen-2.18.0/SettingPresets /tmp/xcodegen_dl/XcodeGen-2.18.0/Sources /tmp/xcodegen_dl/XcodeGen-2.18.0/Tests /tmp/xcodegen_dl/XcodeGen-2.18.0/_config.yml /tmp/xcodegen_dl/XcodeGen-2.18.0/scripts js/ios/vendor/xcodegen
🤖 . 💣 /tmp/xcodegen_dl
✅ XcodeGen: built
```

## error while running target

the logging and childs stderr handling is still very rough, when running paralelisable rules, logs will get mangled for now, I want to address that before merging (if I ever merge)

```
❯ mage xcodeGen
🔨 XcodeGen: building
🤖 . 🏠 /tmp/xcodegen_dl js/ios/vendor/xcodegen
🌏 . ⬇️  https://github.com/yonaskolb/XcodeGen/archive/2.18.0.zip > /tmp/xcodegen_dl/xcodegen.zip
🏃 . ❯ unzip -o -q /tmp/xcodegen_dl/xcodegen.zip -d /tmp/xcodegen_dl
🏃 . ❯ cp -rf --wololo /tmp/xcodegen_dl/XcodeGen-2.18.0/.gitattributes /tmp/xcodegen_dl/XcodeGen-2.18.0/.github /tmp/xcodegen_dl/XcodeGen-2.18.0/.gitignore /tmp/xcodegen_dl/XcodeGen-2.18.0/.swiftformat /tmp/xcodegen_dl/XcodeGen-2.18.0/Assets /tmp/xcodegen_dl/XcodeGen-2.18.0/CHANGELOG.md /tmp/xcodegen_dl/XcodeGen-2.18.0/CONTRIBUTING.md /tmp/xcodegen_dl/XcodeGen-2.18.0/Docs /tmp/xcodegen_dl/XcodeGen-2.18.0/LICENSE /tmp/xcodegen_dl/XcodeGen-2.18.0/Makefile /tmp/xcodegen_dl/XcodeGen-2.18.0/Package.resolved /tmp/xcodegen_dl/XcodeGen-2.18.0/Package.resources /tmp/xcodegen_dl/XcodeGen-2.18.0/Package.swift /tmp/xcodegen_dl/XcodeGen-2.18.0/README.md /tmp/xcodegen_dl/XcodeGen-2.18.0/RELEASE.md /tmp/xcodegen_dl/XcodeGen-2.18.0/SettingPresets /tmp/xcodegen_dl/XcodeGen-2.18.0/Sources /tmp/xcodegen_dl/XcodeGen-2.18.0/Tests /tmp/xcodegen_dl/XcodeGen-2.18.0/_config.yml /tmp/xcodegen_dl/XcodeGen-2.18.0/scripts js/ios/vendor/xcodegen
cp: unrecognized option '--wololo'
Try 'cp --help' for more information.
❌ XcodeGen: error: exit status 1
Error: exit status 1
```

## parralelization

the next one is snippet of `make ios.debug` equivalent after `git clean -xffdq`,
you can see that all those are started imediately in parallel
- xcodegen download
- node modules install
- ruby gems install
- tor libs install

```
❯ mage iosDebug
🏃 . ❯ go mod download
🔨 PbjsFlags: building
🏃 js ❯ go list -m -modfile=../go.mod -mod=mod -f {{.Dir}} github.com/grpc-ecosystem/grpc-gateway
🔨 XcodeGen: building
🤖 . 🏠 /var/folders/dz/0lbcz4t914v9v3ndhvwlc76m0000gp/T/xcodegen_dl js/ios/vendor/xcodegen
🔨 TorVersion: building
🏃 . ❯ go list -mod=readonly -modfile=go.mod -f {{.Version}} -m berty.tech/go-libtor
🌏 . ⬇️  https://github.com/yonaskolb/XcodeGen/archive/2.18.0.zip > /var/folders/dz/0lbcz4t914v9v3ndhvwlc76m0000gp/T/xcodegen_dl/xcodegen.zip
🔨 NodeModules: building
🏃 js ❯ yarn --non-interactive
🔨 RubyGems: building
🏃 js/ios ❯ bundle install
🏃 js ❯ go list -m -modfile=../go.mod -mod=mod -f {{.Dir}} github.com/gogo/protobuf
✅ TorVersion: built
🔨 IOSTor: building
🤖 . 🏠 js/ios/tor-deps /var/folders/dz/0lbcz4t914v9v3ndhvwlc76m0000gp/T//tor-deps/ios
🌏 . ➡️  https://github.com/berty/go-libtor/releases/tag/v1.0.333
(...)
```

## rules list

```
❯ mage         
Targets:
  androidDebug         Build and run debug android app
  androidDebugRe       Force to reinstall the android app
  androidFramework     Build android gomobile framework
  checkGo              Check if go is available
  checkJavaVer         Check that the installed java version matches berty requirements
  checkNpx             Check if npx is available
  checkRubyBundle      Check if bundle is available
  checkSwift           Check if swift is available
  checkTar             Check if tar is available
  checkUnzip           Check if unzip is available
  checkYarn            Check if yarn is available
  frontGen             Build javascript and typescript generated files
  gomobileInit         Initialize gomobile
  iosDebug             Build and run debug ios app
  iosDebugRe           Force to reinstall the ios app
  iosFramework         Build iOS gomobile framework
  iosTor               Fetch tor for iOS
  metroStart           Run metro in foreground (first kill program running on port if any)
  nodeModules          Install node modules
  pbjsFlags            Build common pbjs flags
  pbjsRoot             Build root protobuf javascript
  pbtsRoot             Build root protobuf typescript
  rubyGems             Build ruby gems
  storeTypes           Build javascript store types
  torVersion           Extract tor version from go module
  welshClientsTypes    Build protobuf welsh clients types
  xcWorkspace          Install CocoaPods and build xcworkspace
  xcodeGen             Fetch XcodeGen
  xcodeProj            Build xcodeproj
```

# Try it

all while in berty repo root, switch to this branch and:

## install mage

```
go get -u -d github.com/magefile/mage
cd $GOPATH/src/github.com/magefile/mage
go run bootstrap.go
```

## run metro

`mage metroStart`

## run ios app

`mage iOSDebugRe`

## run android app

`mage androidDebugRe`

## list rules

`mage`

# TODO

## Own

- [ ] handle logs like `go test`
- [ ] add ci cache helpers
- [ ] extract helpers, clean structure
- [ ] cancel faster on failure

## Feedback

- [ ] Check-in `ruby`

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -4,6 +4,7 @@ go 1.14
 
 require (
 	bazil.org/fuse v0.0.0-20200524192727-fb710f7dfd05 // indirect
+	berty.tech/berty v1.0.3
```
Comment (moul) [inline go.mod:7]:
```suggestion
```

File: mage_hashtarget.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
+// +build mage
```
Comment (moul) [inline mage_hashtarget.go:1]: can you move all these go files in a dedicated dir? I suggest `./go/internal/mage`

Comment (moul) [reply mage_hashtarget.go:1]: btw, you can keep the magefile.go at the root, its dependencies should not pollute the root dir

Comment (n0izn0iz) [reply mage_hashtarget.go:1]: I tried at first, will try again

File: go.mod (hunk 1 above)
Comment (n0izn0iz) [reply go.mod:7]: I don't know why these imports get added

**Pull request closed without merging**